src/coprocessor/endpoint.rs

```diff
@@ -150,104 +125,251 @@ impl Context for CopContext {
                         .unwrap();
                 }
             }
-            *this_statistics = Default::default();
+            *this_statistics = StatisticsSummary::default();
         }
-        if !self.request_stats.is_empty() {
-            let mut to_send_stats = HashMap::default();
-            mem::swap(&mut to_send_stats, &mut self.request_stats);
-            if let Err(e) = self.sender.schedule(PdTask::ReadStats {
-                read_stats: to_send_stats,
-            }) {
-                error!("send coprocessor statistics: {:?}", e);
-            };
+        let pd_task = PdTask::ReadStats {
+            read_stats: mem::replace(&mut self.flow_stats, map![]),
+        };
+        if let Err(e) = self.pd_task_sender.schedule(pd_task) {
+            error!("send coprocessor statistics: {:?}", e);
         }
-        self.flush_scan_count();
     }
 }
 
+struct CopContext(RefCell<CopContextInner>);
+
+impl CopContext {
+    fn new(pd_task_sender: FutureScheduler<PdTask>) -> Self {
+        let inner = CopContextInner {
+            select_stats: StatisticsSummary::default(),
+            index_stats: StatisticsSummary::default(),
+            flow_stats: map![],
+            timer: Instant::now_coarse(),
+            timeout: Duration::from_secs(1),
+            pd_task_sender: pd_task_sender,
+        };
+        CopContext(RefCell::new(inner))
+    }
+}
+
+unsafe impl Sync for CopContext {}
```

Comment:
I don't think CopContext is sync.

Reply:
Yes it's not. But `CopContext` is not public. We make it `Sync` for telling compiler.